### PR TITLE
feat: add workflow support for custom operation (drop 2)

### PR DESCRIPTION
**Pull request by Ruadhri17**
## Proposed changes
TODO:

- [x] thin-edge command conversion
- [x] subscribe to command topics for custom operation
- [x] state change handler
- [x] main device system test
- [x] child device system test
- [x] fix bug with `externalID` 
- [x] fix codecov (maybe)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
* #3095 
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (Bravo555) [inline crates/core/c8y_api/src/smartrest/operations.rs:261]: question: these fields are meant to be for `.template` files, but right now it's technically possible for them to be in non-template files as well, right?

**Review by Bravo555 — CHANGES_REQUESTED** [review]
Blockers:
- function that creates symlinks ignores some errors
- memory leak in `From<CumulocitySupportedOperations> for &str`
- typo

Other than that, some redundant branches could use some more terseness. Modified modules could also use some general cleanup, but that can be defered to another PR.

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/operations.rs:330]: Yes, that is leftover after previous solution. I will remove it in follow-up fixup commit

File: tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot
Comment (rina23q) [reply tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot]: It's done. Now `Symlink Should Exist` keyword is available.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:1602]: This `operation_name` is equal to `c8y_operation_name`.

```suggestion
            let ops_file = ops_dir.join(&c8y_operation_name);
```

It means we can simplfy `match_template_with_operation_by_command_name()`. It can return only single value.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:1599]: It took a lot of time to understand why this function works. It seems that the argument (`c8y_operation_name`) is the c8y's operation name (e.g. `c8y_Command`), however, the function checks if the template's command name (e.g. `command`) equals to the c8y operation name.

```rust
    /// Return operation and template name if `command_name` match incoming command
    pub fn match_template_with_operation_by_command_name(
        &self,
        command: &str,
    ) -> Option<(String, Option<String>)> {
        self.templates
            .iter()
            .find(|x| {
                x.command_name()
                    .is_some_and(|command_name| command_name.eq(command))
            })
            .map(|template| (template.name.clone(), template.on_fragment()))
    }
```

But then why it works? I figured it out that the problem deeper, is in `try_convert_data_message()`.

```rust
        match &channel {
            ...
                Channel::CommandMetadata { operation } => {
                    ...
                    OperationType::Custom(c8y_op_name) => {
                        dbg!(&c8y_op_name);
                        self.register_custom_operation(&source, c8y_op_name)
                    }
                }
            ...
        }
```

`OperationType::Custom(c8y_op_name)` holds actually **command** name! Not **c8y_op_name**!
Indeed I added as above, and it prints out `command`.
```
[crates/extensions/c8y_mapper_ext/src/converter.rs:1361:25] c8y_op_name = "command"
```

This needs to be fixed. Here `c8y_op_name` should hold the right value. Then, unfortunately, the function `match_template_with_operation_by_command_name()` needs to be rewrite to check if `template.on_fragment()` is equals to the c8y operation.

**Review by rina23q — CHANGES_REQUESTED** [review]
Unfortunately, there is mistake in converter for custom operation. It needs to be fixed first, then the implementation in this PR should be modified.

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/operations.rs]: fixed by 451bb9146abda49a6cbe646643bf0006349628db

File: crates/common/tedge_utils/src/file.rs
Comment (Ruadhri17) [reply crates/common/tedge_utils/src/file.rs]: fixed by 451bb9146abda49a6cbe646643bf0006349628db

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/operations.rs]: fixed by 451bb9146abda49a6cbe646643bf0006349628db

File: crates/core/c8y_api/src/smartrest/smartrest_serializer.rs
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/smartrest_serializer.rs]: fixed by 24f3203ab59882b6db7b682feb137f017ec74f75

File: crates/core/c8y_api/src/smartrest/smartrest_serializer.rs
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/smartrest_serializer.rs]: fixed by 24f3203ab59882b6db7b682feb137f017ec74f75

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: fixed by 2bedc99af4cc8593b01f143a8d38dab5483954b2

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: After discussion with @rina23q and @albinsuresh, the fix was implemented in 451bb9146abda49a6cbe646643bf0006349628db

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:768]: Unfortunately, this function is called from two places: `parse_json_custom_operation_topic()` and `process_json_over_mqtt()` as both `C8yDeviceControlTopic` and custom topics can handle this kind of operation. In the latter one all that parameters are extracted so in order to not create two functions for those case or two instances of those parameters I decided to just pass those three you mentioned.

File: crates/extensions/c8y_mapper_ext/src/operations/handlers/custom_operation.rs
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/operations/handlers/custom_operation.rs]: fixed by 30c84b15e37d73357842338844ed8fe2fee64498

File: crates/common/tedge_utils/src/file.rs
Comment (Bravo555) [reply crates/common/tedge_utils/src/file.rs]: > should return [EEXIST](https://www.man7.org/linux/man-pages/man2/symlink.2.html#ERRORS) if symlink already exists, and if so, we can just ignore it.

I may have mislead you a little there. After some more checking, `EEXIST` will also be returned if a symlink exists but points to another file entirely, so if we just continue, we may end up with a wrong path.
At this point we can either return an error or overwrite, depending on if we can recover ourselves and not screw up any user data, which given that only we create these symlinks, may be the case. Or you can just return the `EEXIST` error, ultimately I'm also not sure how to handle this optimally.

Comment (Ruadhri17) [reply crates/common/tedge_utils/src/file.rs]: In that case we can use `read_link()` to check if those pathes are matching. If not we can return error to override symlink.

**Review by rina23q — COMMENTED** [review]
A bit more work required

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (albinsuresh) [inline crates/core/c8y_api/src/smartrest/operations.rs:145]:
```suggestion
    pub fn get_operation_template_by_command_name(&self, command: &str) -> Option<String> {
```

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (albinsuresh) [inline crates/core/c8y_api/src/smartrest/operations.rs:153]: This makes me wonder if the `on_fragment` filed is redundant, when the name of the operation file should have been be the same. Are there any cases in C8y where the operation name and this fragment name could be different?

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/operations.rs:145]: I would go with `get_operation_name_by_command_name` or `get_operation_fragment_by_command_name`

Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/operations.rs]: resolved by 1f913a33c1d9e6e87664d5e184192e6f512b67ce

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/operations.rs]: resolved by 1f913a33c1d9e6e87664d5e184192e6f512b67ce

File: crates/extensions/c8y_mapper_ext/src/tests.rs
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/tests.rs]: resolved by 3d322e43ad1493e0f71bec602c69f787c3d1741a

File: crates/extensions/c8y_mapper_ext/src/tests.rs
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/tests.rs]: resolved by 3d322e43ad1493e0f71bec602c69f787c3d1741a

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/operations.rs]: resolved by 1f913a33c1d9e6e87664d5e184192e6f512b67ce

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/operations.rs]: resolved

File: crates/extensions/c8y_mapper_ext/src/tests.rs
Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/tests.rs:1573]: As discussed offline, automatically publishing the 114 message on any updates to the operations directory was limited only to main device addressing https://github.com/thin-edge/thin-edge.io/issues/2409

File: tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot
Comment (Bravo555) [inline tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot:53]: question: Is my understanding correct that this line checks if `c8y_TakePicture` is supported by the parent device, not the child device, so it will always be true because we register `c8y_TakePicture` in the suite setup?

I see that we send `114,...,c8y_TakePicture` on MQTT topic `c8y/s/us/TST_augment_connected_sesame:device:TST_bark_randomized_exit`, so we should be checking that it's the child device that supports the operation, right?

Given that both devices support that operation, I don't understand how it ends up on the child device and not the main device.

File: tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot
Comment (Bravo555) [inline tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot:77]: nit: typo

Comment (Ruadhri17) [reply tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot]: fixed by a531772c004475ae768f6b978b8b37a2a35d8de0

File: tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot
Comment (Bravo555) [reply tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot:53]: Had a call about this with @Ruadhri17

> Given that both devices support that operation, why does incoming message on c8y/devicecontrol/notifications has externalSource of child device and it ends up being executed on the child device and not the main device?

`Cumulocity.Create Operation` seems to set `externalSource` to the most recent child device that supports the operation, but ultimately which device receives an operation can be decided by the user, using e.g. `go-c8y-cli`. So my concern was just due to my unfamiliarity with this Cumulocity feature and not anything related to thin-edge.

> I see that we send 114,...,c8y_TakePicture on MQTT topic c8y/s/us/TST_augment_connected_sesame:device:TST_bark_randomized_exit, so we should be checking that it's the child device that supports the operation, right?

At the end of the test we verify that the command is executed by the child device, so these previous checks that device exists and device supports the operation may be redundant, but most importantly they are slightly misleading because wrong device is checked.

These checks should be fixed to use a correct device, but they could also just be removed since they're redundant.

Comment (albinsuresh) [reply tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot:53]: > Cumulocity.Create Operation seems to set externalSource to the most recent child device that supports the operation, but ultimately which device receives an operation can be decided by the user, using e.g. go-c8y-cli.

Actually, it's the `Cumulocity.Device Should Exist    ${CHILD_XID}` call that changes the context to that device and then all subsequent C8y operations would be done on that device. Yes, implicitly changing the context like that was a bit confusing for me also, when I initially started using this library and have gotten used to it over time. I agree that it would have been better if it was clearer from the API name itself.

**Review by albinsuresh — COMMENTED** [review]
Happy to approve once https://github.com/thin-edge/thin-edge.io/pull/3225#discussion_r1848879538 is also resolved. If that `on_fragment` key is redundant (always same as the operation name itself), we can get rid of it.

File: crates/core/c8y_api/src/smartrest/operations.rs
Comment (reubenmiller) [reply crates/core/c8y_api/src/smartrest/operations.rs:145]: I wouldn't that the fragment and the name of the operation are the same. These are custom operations which means that the customer is free to configure them however they want, so being more explicit would be safer here with the downside of a bit of boilerplate.

Comment (albinsuresh) [reply crates/core/c8y_api/src/smartrest/operations.rs:145]: @Ruadhri17 if you're also wondering why any user would want to declare one name in the supported operations list (the operation file name) and use an entirely different name for his operation fragment, here is the use-case: The declared supported operation name is typically used to enable/disable UI components in C8y. The developer of that UI component dictates the expected operation name. And it is a possibility that the custom operation developer may not be able to (or don't want to) reuse the same name for his operation as well. In such cases, this design would help.

File: tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot
Comment (Bravo555) [reply tests/RobotFramework/tests/cumulocity/custom_operation/custom_operation_workflow/custom_operation_workflow.robot:53]: > Actually, it's the Cumulocity.Device Should Exist ${CHILD_XID} call that changes the context to that device and then all subsequent C8y operations would be done on that device

Ah, that explains it, and as there's no mention of it in the documentation of the keyword, the implicit change of the context is quite unexpected indeed.

**Review by Bravo555 — APPROVED** [review]
https://github.com/thin-edge/thin-edge.io/pull/3225/files#r1851457754

Thanks to Albin's comment I now know that the checks are indeed correct, so I don't have any other blockers. 

The code and the test suite could use some more documentation comments, to make it easier why we do the things we do without retracing the original tickets, but I don't want to block this PR anymore.

File: crates/extensions/c8y_mapper_ext/src/operations/convert.rs
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/operations/convert.rs:425]: This can be improved using the same logic as for spawning a sub-operation using [`StateExcerpt`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_api/src/workflow/state.rs#L540)

```
let excerpt = StateExcerpt::PathExpr(custom_handler.tedge_cmd_payload() );
let payload = excerpt.extract_value_from(state);
```

The main benefit is that this will help to compatible with workflow definition and enable feature like building the tedge operation payload as a *combination* of several fragments of the c8y operation payload (as done [here](https://thin-edge.github.io/thin-edge.io/references/agent/operation-workflow/#sub-operation-execution):

```
command_payload.x = "some static value"
command_payload.y = "${.payload.y}"
command_payload.z = "${.payload.z}"
```

Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/operations/convert.rs:425]: I'm okay to merge this PR without this refinement.

However, we need to agree on the property names as discussed [here](https://github.com/thin-edge/thin-edge.io/issues/3095#issuecomment-2473876874) and [here](https://github.com/thin-edge/thin-edge.io/issues/3095#issuecomment-2490426004). @reubenmiller what's your point of view of these property names?

Comment (reubenmiller) [reply crates/extensions/c8y_mapper_ext/src/operations/convert.rs:425]: Ok, I've added my suggestion which is merging the two concepts. https://github.com/thin-edge/thin-edge.io/issues/3095#issuecomment-2493486901

**Review by didier-wenzek — APPROVED** [review]
Approved. Thank you for taking the time to address all the comments.

**Review by albinsuresh — APPROVED** [review]
LGTM

**Pull request merged**
